vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py

```diff
@@ -1738,90 +1738,113 @@ def save_kv_to_host(self, metadata: NixlConnectorMetadata):
-            cache_list = cache_or_caches if split_k_and_v else [cache_or_caches]
-            for cache in cache_list:
-                blocks_to_update = cache.index_select(0, indices)
-                permuted_blocks = blocks_to_update.reshape(src_shape).permute(
-                    *inv_order
-                )
-                cache.index_copy_(0, indices, permuted_blocks)
-
-    def blocksize_post_process(self, block_ids_per_ratio: dict[int, list[list[int]]]):
-        def _process_local_gt_remote(blocks_to_update, block_size_ratio):
+    def post_process_device_kv_on_receive(
+        self,
+        block_size_ratio: int,
+        enable_permute_local_kv: bool,
+        block_ids_list: list[list[int]],
+    ):
+        def _kv_postprocess_blksize(cache, indices, block_size_ratio):
+            """
+            Transforms the layout of received KV cache blocks to the local block_size.
+            (Only works for local blocksize > remote blocksize)
+
+            example:
+            local blocksize = 16 tokens, remote blocksize = 4 tokens
+            local block[0] = remote block[0, 1, 2, 3]
+            remote is |h0-b0|h1-b0|h2-b0|h3-b0|h0-b1|h1-b1|h2-b1|h3-b1|...
+            local is  |h0-b0..................|h1-b0..................|...
+            permute is to:
+            1. view => view remote as n_blocks * remote_shape(H,remoteN,D)
+            2. permute => (H, nblocks, remoteN, D)
+            3. flatten => (H, localN, D)
+            """
+            blocks_to_update = cache.index_select(0, indices)
+            # use physical order
+            blocks_to_update = blocks_to_update.permute(0, 2, 1, 3)
             n_kv_heads, block_size, head_size = blocks_to_update.shape[1:]
             remote_block_size = block_size // block_size_ratio
             n_blocks = block_size_ratio
-            # actual permute is to convert
-            # for local blocksize > remote blocksize
-            # ex: local blocksize = 16 tokens, remote blocksize = 4 tokens
-            # local block[0] = remote block[0, 1, 2, 3]
-            # remote is |h0-b0|h1-b0|h2-b0|h3-b0|h0-b1|h1-b1|h2-b1|h3-b1|...
-            # local is  |h0-b0..................|h1-b0..................|...
-            # permute is to:
-            # 1. view => view remote as n_blocks * remote_shape(H,remoteN,D)
-            # 2. permute => (H, nblocks, remoteN, D)
-            # 3. flatten => (H, localN, D)
+
             permuted_blocks = (
                 blocks_to_update.reshape(
                     -1, n_blocks, n_kv_heads, remote_block_size, head_size
                 )
                 .permute(0, 2, 1, 3, 4)
                 .flatten(2, 3)
             )
-            return permuted_blocks
+            permuted_blocks = permuted_blocks.permute(0, 2, 1, 3)
+            cache.index_copy_(0, indices, permuted_blocks)
+
+        def _kv_postprocess_layout(cache, indices):
+            """Transforms the layout of received KV cache blocks to the local format.
```

Comment:
what about moving these 3 refactored function into utils.py to reclaim some space?

Reply:
I mean the `_kv_postprocess_*` functions

Reply:
Ok, will do

Reply:
done